tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c

```diff
@@ -0,0 +1,315 @@
+	ASSERT_EQUAL(buf->len, length);
+	ASSERT_EQUAL(length, sizeof(uint8_t) + strlen(txt_msg));
+	ASSERT_EQUAL(BT_DATA_MESH_MESSAGE, net_buf_simple_pull_u8(buf));
+
+	char *data = net_buf_simple_pull_mem(buf, strlen(txt_msg));
+
+	ASSERT_EQUAL(0, memcmp(txt_msg, data, strlen(txt_msg)));
+
+	cnt++;
+
+	if (cnt > 1) {
+		delta = k_uptime_delta(&buff_timestamp);
+
+		ASSERT_TRUE(delta >= xmit_param->interval &&
+			delta < (xmit_param->interval + xmit_param->interval / 2));
+	}
+
+	LOG_INF("rx: %s +%d ms", txt_msg, delta);
+
+	buff_timestamp = timestamp;
+
+	if (cnt > xmit_param->retr) {
+		cnt = 0;
+		delta = 0;
+		buff_timestamp = 0;
+		k_sem_give(&observer_sem);
+	}
+}
+
+static void rx_xmit_adv(void)
+{
+	struct bt_le_scan_param scan_param = {
+		.type       = BT_HCI_LE_SCAN_PASSIVE,
+		.options    = BT_LE_SCAN_OPT_NONE,
+		.interval   = BT_MESH_ADV_SCAN_UNIT(1000),
+		.window     = BT_MESH_ADV_SCAN_UNIT(1000)
+	};
+	int err;
+
+	k_sem_init(&observer_sem, 0, 1);
+	err = bt_le_scan_start(&scan_param, xmit_scan_cb);
+
+	if (err && err != -EALREADY) {
+		FAIL("starting scan failed (err %d)", err);
+	}
+
+	err = k_sem_take(&observer_sem,
+		K_MSEC(((xmit_param->retr + 1) * xmit_param->interval) + 10));
+	if (err) {
+		FAIL("Didn't receive adv in time (err: %d)", err);
+	}
+
+	err = bt_le_scan_stop();
+	if (err && err != -EALREADY) {
+		FAIL("stopping scan failed (err %d)", err);
+	}
+}
+
+static void test_tx_cb(void)
+{
+	char cb_tst_msg[strlen(cb_msg)];
+	struct net_buf *buf;
+	int err;
+
+	bt_init();
+	adv_init();
+
+	buf = bt_mesh_adv_create(BT_MESH_ADV_DATA, BT_MESH_LOCAL_ADV,
+			BT_MESH_TRANSMIT(2, 20), K_NO_WAIT);
+	if (!buf) {
+		FAIL("Out of buffers");
+	}
+
+	send_cb.start = single_start_cb;
+	send_cb.end = single_end_cb;
+
+	k_sem_init(&observer_sem, 0, 1);
+	net_buf_add_mem(buf, txt_msg, strlen(txt_msg));
+	memcpy(cb_tst_msg, cb_msg, strlen(cb_msg));
+	seq_checker = 0;
+	tx_timestamp = k_uptime_get();
+	bt_mesh_adv_send(buf, &send_cb, cb_tst_msg);
+	net_buf_unref(buf);
+
+	err = k_sem_take(&observer_sem, K_SECONDS(1));
+	if (err) {
+		FAIL("Didn't call end tx cb.");
+	}
+
+	PASS();
+}
+
+static void test_rx_xmit(void)
+{
+	struct bt_mesh_test_adv param = {
+		.retr = 2,
+		.interval = 20
+	};
+
+	bt_init();
+	xmit_param = &param;
+	rx_xmit_adv();
+	PASS();
+}
+
+static void test_tx_cb_seq(void)
+{
+	struct net_buf *buf[CONFIG_BT_MESH_ADV_BUF_COUNT];
+	struct net_buf *dummy_buf;
+	int buf_id[CONFIG_BT_MESH_ADV_BUF_COUNT];
+	int err;
+
+	bt_init();
+	adv_init();
+	k_sem_init(&observer_sem, 0, 1);
+
+	for (int i = 0; i < CONFIG_BT_MESH_ADV_BUF_COUNT; i++) {
+		buf[i] = bt_mesh_adv_create(BT_MESH_ADV_DATA, BT_MESH_LOCAL_ADV,
+				BT_MESH_TRANSMIT(2, 20), K_NO_WAIT);
+		if (!buf[i]) {
+			FAIL("Out of buffers");
+		}
+	}
+
+	dummy_buf = bt_mesh_adv_create(BT_MESH_ADV_DATA, BT_MESH_LOCAL_ADV,
+			BT_MESH_TRANSMIT(2, 20), K_NO_WAIT);
+	if (!!dummy_buf) {
+		FAIL("Unexpected extra buffer");
+	}
+
+	send_cb.start = NULL;
```

Comment:
Please add some comments before each test block to make it easier understand the code when we need to look at it later if it gets broken, e.g.:
```suggestion
	/* Test that adv buf is unreferenced before end callback is called. */
	send_cb.start = NULL;
```